### PR TITLE
Integrate Comments to Equations Tutorial Basics

### PR DESCRIPTION
**Pull request by thomas-lamiaux**
Integrate Comments to Equations Tutorial Basics

- [x] l16: no-confusion/subterm -> these are probably quite obscure for beginners (at least no-confusion), maybe you should explain more or drop it?
- [x] l102: provided they can be inferred. For instance, above we wrote…
- [x] l105: We can use underscores…
- [x] l108: has constructors it is…
- [x] l113: why the parentheses around the underscore? (here and other places)
- [x] l135 the [Fixpoint] one -> the one provided by the [Fixpoint] command and Coq's native pattern-matching.
- [x] l151: acceptable -> can be used interchangeably.
- [X] l158: (n : nat ) -> (n : nat)
- [X] l179: c.f -> c.f.
- [x] l184: finish the proof, to show that you need to case-split on [n]?
```
case n as [|n'].
all: autorewrite with nth_option.
all: reflexivity.
Abort.
```
Also, at that point you pretend that this is about computation, but
`autorewrite` is propositional reasoning. That is, you would be able to add the new
derived equation to the `nth_option` hintbase:
```
Lemma nth_option_nil : forall {A} n, nth_option n (nil (A := A)) = None.
Proof.
  intros.
  case n ; now autorewrite with nth_option.
Qed.

Hint Rewrite @nth_option_nil : nth_option.
```
And now the first `autorewrite` goes through.
I'm not sure it is something you want to flash here, or even later, but it might confuse learners that you talk about computation…
- [X] l187: on [a, …, a] -> on [a1, …, an]
- [X] l194: is it intentional that `map` is filled, but not the other two? (It would make sense to me that you want to give at least one example, but force people to look for themselves in the other too, until we get a better set-up for exercises)
- [X] l219: the functions -> functions, can not -> cannot (here and in other places)
- [X] l245-246: `Arguments` is probably not precise enough to be able to say what one would want to say ("simplify exactly when the list of arguments matches one of the patterns of the definition"), but your proposed version is *very aggressive*. Usually, for `app` I'd rather do `Arguments app !_ /.` (which means "unfold `app` when it is applied to at least one argument and that argument is a constructor). There's also the `simpl nomatch` invocation, which I've never used but seems relevant…
- [X] l266: mention how each equalities are named (it always takes me forever to find it again)? Although using the database is usually the sensible way to go, sometimes its useful to be able to summon the exact equality…
- [X] l266: declares -> proves, and again below
- [X] l266: in a database -> collected in a hint database
- [x] l271: when doing in proof as compared to cbn the user … -> in proofs compared to cbn, as the user…
Also, this is misleading: by using `cbn [f1 … fn]` one can obtain a similar behaviour. So it's maybe best not to mention this point at all.
- [X] l277: uses -> use
- [X] l278: We can similarly -> make this explicitly an exercise?
- [X] l295: detruct
- [X] l300: a more idiomatic way would be `induction l in n |- *` (although I know in the end you want to say "functional induction is better" anyway).
- [x] l309: I'd say the important aspect is not that the inductive types can get more complicated (although they certainly can), but that the functions might not be simply following the structure of the types, without even talking about [with]/[where] clauses? Like in the most trivial example
```
Equations half (n : nat) : nat :=
  half 0 := 0 ;
  half 1 := 0 ;
  half (S (S n)) := S (half n).
```
- [X] l311-313: maybe mention that this is called functional induction? Attempt:
> Consequently, for each definition, [Equations] derives a functional induction
> principle: this is an induction principle that follows the structure of the
> function, including deep pattern-matching, [with] and [where] clauses,
> correct induction hypotheses and generalisation as in the above example, etc.
> This induction principle can be accessed by using the [funelim] tactic.
- [X] l318: do your induction -> do the induction on (or better? induct on)
- [X] l320: let's -> let us (here and other places)
- [X] l352: to automatically simply by when -> use for automatic simplification
- [X] l356: to make it automatic -> this way
- [X] l358: by the syntax -> with the following syntax
Maybe make it clear that this is not about Equations per se, but simply using the
general rewriting hint mechanism of Coq?
- [X] l360: name_function -> function_name
- [X] l362: provides us with -> provides a
- [X] l363-364: I would rephrase to
> However, note that this mechanism is expressive enough to be
> non-terminating if the wrong lemmas are added to the database.
- [x] Add something as to the fact that the default lemmas are always safe, ie
confluent and terminating?
- [X] In general, I would replace "please note" with simply "note"
- [X] l400: what is "it"?
- [X] l404: I would rephrase to "… happens in such proofs by functional induction that
after simplification we get a lot of uninteresting cases, that we would like to
deal with in an automatic but controlled way."
- [x] l409: a proof search by -> a proof search using
"a particular instance" is rather vague, do you mean it uses `try typeclasses eauto with …` using a given set of hints? Then you should say so, mention what the databases are and what they do.
- [X] l447: you are doing -> you want to do?
(Good examples, especially `app_nth2`, I would never have thought about doing
`funelim (nth_option n l)` despite `nth_option n l` appearing nowhere in my goal!)
- [x] It is a bit weird to mention the useful lemmas but not import the corresponding files. And you might suggest using `lia` instead, which gets rid of these arithmetic goals without needing to find the right lemma.
- [x] Maybe you should also mention somewhere that you can use `simp in H` and `simp in *` to also simplify in the hypotheses? This is useful in these exercises…
- [x] Before 2: you do not have any example of a deep pattern-matching, as in the `half` example above. Maybe it would be useful to have something like this to show the power of `funelim`?
- [X] l479: Agda -> dependently-typed programming community? (Ie Idris has it too for instance)
- [X] l498: verify -> satisfy
- [X] l532: |with] -> [with]
- [X] l537: makes a subclause… -> does a case-split corresponding to the [with] clause?
- [X] l539: the sentence is strange, attempt:
> For [filter], in the [p a = true] case, a [filter_clause_1] appears, and
> similarly in the [false] case.
- [X] l541: [Equations] -> [Equations]' 
- [X] l541-544: I wonder whether this is an opportunity for polishing on the Equations side…
- [X] l568: requiring for the input … different than -> requiring the input … different from
- [X] l566: They can also be useful to… -> Finally, [with] clauses can be used to…
- [X] l571: as -> has
- [X] l604: your function -> your functions
- [X] l625: using … -> using the usual [Equations] syntax/using [Equations]' usual syntax
- [X] l651: maybe add some `About rev_acc_elim.` at this point so that we get to see what its type is?
- [X] l656: `exact … ` -> `1: exact …` (I'm picky about goal selectors :p) (and below)
- [X] l665: automatised -> automated
- [X] l676: now you use `simp in` without having introduced it before!

### CODE REVIEW

**Comment by thomas-lamiaux** [comment]
@MevenBertrand can you elaborate on:
- what is the point of reasoning with rewrite than cbn according to you ?
- what do you mean by 
> Add something as to the fact that the default lemmas are always safe, ie confluent and terminating?
- I wanted to be more specific about simp, but I have no idea what subterm and Below are

Also the behaviour of funelim should change soon, c.f. this comment https://github.com/Zimmi48/platform-docs/pull/1#issuecomment-2098810034

**Comment by thomas-lamiaux** [comment]
I could use help here https://github.com/Zimmi48/platform-docs/pull/19/files#diff-c729063ee0f5b61fd3a9901b51f3de77b92d8718093c0d8d894f060835c765f6R450

**Comment by thomas-lamiaux** [comment]
Refactored 1.2 to adpt to the comments

**Comment by MevenBertrand** [comment]
> what is the point of reasoning with rewrite than cbn according to you ?
I'm not sure I understand the question ^^'

> I wanted to be more specific about simp, but I have no idea what subterm and Below are
These are hint databases used to respectively solve goals related to the subterm relation (whenever you `Derive Subterm …` various lemmas about that subterm relation are generated and added to the hint database), and well-foundedness assumption. If you want to know what is in a hint database, `Print HintDb …` is your friend :)

**Comment by thomas-lamiaux** [comment]
@MevenBertrand Could you check section 1.2 and tell me if you are satisfied by the modifications ?

File: src/Tutorial_Equations_basics.v [hunk 1]
```diff
@@ -49,17 +53,18 @@ Arguments to_fill {_}.
 
 (** * 1. Basic definitions and reasoning
 
-  Let's start by importing the package:
+  let us start by importing the package:
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:56]: Capitalization (I guess this was grep ? :p)

File: src/Tutorial_Equations_basics.v [hunk 2]
```diff
@@ -49,17 +53,18 @@ Arguments to_fill {_}.
...
 *)
 
+From Coq Require Import PeanoNat.
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:59]: I *believe* the standard here is `Import Arith.` (a bundle file contain all things about arithmetic, including `PeanoNat`)

File: src/Tutorial_Equations_basics.v [hunk 3]
```diff
@@ -148,25 +153,26 @@ Equations app'' {A} (l l' : list A) : list A :=
...
+nth_option 0 []     := None ;
 nth_option 0 (a::l) := Some a ;
-nth_option (S n) nil := None ;
+nth_option (S n) []     := None ;
 nth_option (S n) (a::l) := nth_option n l.
 
 (** However, be careful that as for definitions using [Fixpoint], the order
-    of matching matters : it produces term that are equal but with
-    different computation rules.
+    of matching matters: it produces terms that are observationally equal but
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:171]: Observational (equality) is a loaded word nowadays. I would probably keep only "equal" (or "propositionally equal", but that might throw off people who do not know about meta-theory shenanigans).

File: src/Tutorial_Equations_basics.v [hunk 4]
```diff
@@ -148,25 +153,26 @@ Equations app'' {A} (l l' : list A) : list A :=
...
+    with different computation rules.
     For instance, if we pattern match on [l] first, we get a function
     [nth_option'] that is equal to [nth_option] but computes differently
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:174]: Technically, it's only pointwise equal, ie you can't prove `nth_option = nth_option'` without funext… Maybe say
> For instance, if we pattern match on [l] first, we get a function [nth_option'] for which we can prove [forall n l, nth_option n l = nth_option' l n], but which computes differently, as can be seen below:

File: src/Tutorial_Equations_basics.v [hunk 5]
```diff
@@ -176,24 +182,34 @@ nth_option' (a::l) (S n) := nth_option' l n.
...
-(** As exercices, you can try to define:
-    - A function [map f l] that applies [f] pointwise, on [ [a, ... ,a] ] it
+(** [Equations] also supports nested pattern matching.
+    For instance, we can define a function swapping lists of pairs
+    by matching pairs at the same as we match the list.
+*)
+
+Equations swap_list_pair {A B} (l : list (A * B)) : list (B * A) :=
+swap_list_pair [] := [];
+swap_list_pair ((a , b)::l) := (b,a)::(swap_list_pair l).
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:201]: :+1:

File: src/Tutorial_Equations_basics.v [hunk 6]
```diff
@@ -238,61 +256,84 @@ Proof.
...
 
-(** If one wishes to recover unfolding, it is possible on a case by case
+(** If one really wishes to recover unfolding, it is possible though on a case by case
     basis using the option [Global Transparent f] as shown below.
     It is also possible to set this option globally with [Set Equations
     Transparent], but it is not recommended to casual users.
-    To recover simplification by [simpl] and [cbn], one needs to additionally
-    use the command [Arguments name_function /].
+    To recover simplification by [simpl] and [cbn], one can additionally
+    use the command [Arguments function_name !_ /. ].
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:264]: Careful that this only works if the function has one arguments,  if you have two you'll probably want `Arguments !_ !_ /.` Not sure how to concretely make the presentation better, though… Maybe this one is ok at this point, this is a side-matter anyway.

File: src/Tutorial_Equations_basics.v [hunk 7]
```diff
@@ -238,61 +256,84 @@ Proof.
...
-  Arguments f4 /. cbn.
+  Arguments f4 !_ /. cbn.
 Abort.
 
 (** Rather than using [cbn] or [simpl], for each function [f], [Equations]
-    internally declares an equality for each case of the pattern matching
-    defining [f], in a database named [f].
-    For instance, for [app], it declares [app [] l' = l'] and
+    internally proves an equality for each case of the pattern matching
+    defining [f], and declares them in a hint database named [f].
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:286]: declares it

File: src/Tutorial_Equations_basics.v [hunk 8]
```diff
@@ -238,61 +256,84 @@ Proof.
...
-    It is then possible to simply simplify by the equations associated
-    to function [f1 ... fn] using the tactic [autorewrite with f1 ... fn].
-    This provide a better control of unfolding when doing in proofs as
-    compared to cbn the user can choose which functions [f1 ... fn],
-    they wants to simplify.
+    The equalities are named using the scheme "function_name_equation_n", and you
+    can check them out using the command [Print Rewrite HintDb function_name]:
+*)
+
+Print Rewrite HintDb app.
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:293]: :+1:

File: src/Tutorial_Equations_basics.v [hunk 9]
```diff
@@ -238,61 +256,84 @@ Proof.
...
-    definition of [app] by [induction l].
+Lemma app_assoc {A} (l1 l2 l3 : list A) : (l1 ++ l2) ++ l3 = l1 ++ (l2 ++ l3).
+Proof.
+  induction l1. all: autorewrite with app.
+  - reflexivity.
+  - rewrite IHl1. reflexivity.
+Qed.
+
+(** This provide a fine control of unfolding as it simplifies only by the defining
+    equations, and will not unfold nothing else, while leaving the possibility to
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:320]: not unfold anything else

File: src/Tutorial_Equations_basics.v [hunk 10]
```diff
@@ -305,36 +346,71 @@ Proof.
...
-    function, there is no need to do it by hand.
-    Moreover, it does so directly with the right induction hypothesis, there
-    is no need to generalise the induction hypothesis as done by
-    reverting [n] in the proof of [nth_eq] above.
-    To use it, it suffices to write [funelim (name_function a1 ... an)]
-    where [a1 ... an] are the arguments you want to do your induction on.
+(** On real life examples, reproducing the patterns by hand with the good
+    induction hypotheses can quickly get tedious, if not challenging.
+    Inductive types and patterns can quickly get complicated.
+    Moreover, function may actually not even be defined following the natural
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:352]: functions
not sure "natural structure" makes sense, I would just keep structure.

File: src/Tutorial_Equations_basics.v [hunk 11]
```diff
@@ -305,36 +346,71 @@ Proof.
...
+half 1 := 0 ;
+half (S (S n)) := S (half n).
+
+Equations mod2 (n : nat) : nat :=
+mod2 0 := 0;
+mod2 1 := 1;
+mod2 (S (S n)) := mod2 n.
+
+(** If we try to naively prove a property [P] about [half] or [mod2]
+    by double induction on [n], we actually quickly get stuck.
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:370]: not sure "double induction" is a good term, I'd say "by using the standard induction principle for `nat`" or something similar.

File: src/Tutorial_Equations_basics.v [hunk 12]
```diff
@@ -305,36 +346,71 @@ Proof.
...
-  funelim (app l1 l2). all : autorewrite with app.
-  - reflexivity.
-  - rewrite H. reflexivity.
-Qed.
+  induction n. 1: exact H0.
+  induction n. 1: exact H1.
+  (* We are stuck *)
+Abort.
+
+(** This issue arise more generally as soon as we start defining functions using
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:384]: arises

File: src/Tutorial_Equations_basics.v [hunk 13]
```diff
@@ -305,36 +346,71 @@ Proof.
...
+    more advanced notions of matching like [with] and [where] clauses (c.f 2. and 3.),
+    or when defining functions using well-founded recursion.
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:386]: Mention the second tutorial?

File: src/Tutorial_Equations_basics.v [hunk 14]
```diff
@@ -305,36 +346,71 @@ Proof.
...
-  - reflexivity.
-  - rewrite H. reflexivity.
-Qed.
+  induction n. 1: exact H0.
+  induction n. 1: exact H1.
+  (* We are stuck *)
+Abort.
+
+(** This issue arise more generally as soon as we start defining functions using
+    more advanced notions of matching like [with] and [where] clauses (c.f 2. and 3.),
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:385]: c.f.

File: src/Tutorial_Equations_basics.v [hunk 15]
```diff
@@ -305,36 +346,71 @@ Proof.
...
+    or when defining functions using well-founded recursion.
+
+    Consequently, to help us reason about complex functions, for each definition,
+    [Equations] derives a functional induction principle.
+    This is an induction principle that follows the structure of the
+    function, including deep pattern-matching, [with] and [where] clauses,
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:391]: Above you used "nested pattern matching", it's probably good to be consistent

File: src/Tutorial_Equations_basics.v [hunk 16]
```diff
@@ -346,24 +422,87 @@ Proof.
...
+Definition nth_eq {A} (l : list A) (n : nat) : nth_option n l = nth_option' l n.
+Proof.
+  funelim (nth_option n l); simp nth_option nth_option'.
+  all : reflexivity.
+Abort.
+
+(** As you can see, by default [simp] does not try to prove goals that hold
+    by definition, like [None = None].
+    If you wish for [simp] to do so, or to try any other tactic, you need to add
+    it as a hint to one of the hint databse used by [simp].
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:470]: databases

File: src/Tutorial_Equations_basics.v [hunk 17]
```diff
@@ -395,45 +537,21 @@ Qed.
...
-    As examples, let's prove again [nth_eq] and [rev_rev_acc]:
-*)
-
-Definition nth_eq {A} (l : list A) (n : nat) : nth_option n l = nth_option' l n.
-Proof.
-  funelim (nth_option n l). all: simp nth_option nth_option'.
-  all : reflexivity.
-Abort.
-
+(** It actually enables to greatly automatise proofs using [simp] *)
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:546]: automatise -> automate

File: src/Tutorial_Equations_basics.v [hunk 18]
```diff
@@ -495,7 +615,7 @@ Admitted.
     predicate [p].
     In the case [cons a l], the [with] clause enables us to compute
     the intermediate value [p a] and to check whether it is [true] or [false],
-    and hence if [a] verify the predicate [p] and should be kept or not:
+    and hence if [a] satisfy the predicate [p] and should be kept or not:
```
Comment (MevenBertrand) [inline src/Tutorial_Equations_basics.v:618]: satisfies

**Review by MevenBertrand — COMMENTED** [review]
Very nice addition of the examples for nested pattern-matching. Some more typo catches, but otherwise, lgtm :)

File: src/Tutorial_Equations_basics.v (hunk 6 above)
Comment (thomas-lamiaux) [reply src/Tutorial_Equations_basics.v:264]: It works if there is at least one argument, so it is an acceptable middle ground

File: src/Tutorial_Equations_basics.v (hunk 11 above)
Comment (thomas-lamiaux) [reply src/Tutorial_Equations_basics.v:370]: I chose "repeated induction on n"

**Comment by thomas-lamiaux** [comment]
Merging now, thanks for the insight @MevenBertrand  !
